### DISCUSSION TITLE
Endangered and undergoal chunks not replicated

### DISCUSSION THREAD

**Discussion by OE2WNL** [moosefs/moosefs#502] [General]
Hi!

I'm using moosefs with 8 Odroid HC2 each with a 8TB drive and mfsmaster as HyperV VM.
MFS: Version 3.0.116 
OS: Debian Buster
Size: 57TB
Freespace on each 8TB HDD is abt. 1TB.

I only switch the whole cluster on when I want to backup my stuff. This worked since some years.

Since I switched it on last week it now shows undergoal/endangered chunks and it will not duplicate the chunks even after several days.
I have at least Goal of 2.

As a test I assigned a goal of 4 to a test text file but this file does not get 4 copies, still sitting on 2.

Master and Chunkservers doing nothing, no duplication no rebalancing happens.
As you see in the screenshot even the "ready to be removed" chunks sitting there and waiting to be processed but nothing happens.
What can I do to get the chunks to duplicate.

Greetings
Wolfgang

Appendix:
CGI-Screenshot:
https://share.derwebwolf.net/s/iLom03gBC8Y1wma

master log shows:
`
Oct 13 12:25:33 mfsmaster mfsmaster[441]: (192.168.64.226:9422 -> 192.168.64.227:9422) chunk: 00000000001F3123 replication status: CRC error

Oct 13 12:41:28 mfsmaster mfsmaster[441]: (192.168.64.225:9422 -> 192.168.64.227:9422) chunk: 000000000003A88B replication status: CRC error

Oct 13 12:44:19 mfsmaster mfsmaster[441]: (192.168.64.226:9422 -> 192.168.64.227:9422) chunk: 0000000000045F6C replication status: CRC error
...
`

**Comment by chogata** [comment]
This is a bit strange. If those chunk copies were indeed invalid (had wrong CRC), they would have been marked invalid a long time ago and your files would become missing. But if you have undergoal files with CRC errors for a couple of days, the problem is somewhere else.
Another strange thing is that the destination server for all the unsuccessful replication tries is always the same (192.168.64.227:9422). Check logs on this server - are there any error messages there? Especially something that says "replicator received data with wrong checksum"? Are there any CRC errors in logs on all the other chunkservers?

**Comment by OE2WNL** [comment]
Hi!

Thanks for your reply!

In the meantime I now did stop the mfschunkserver process on the "64.227" (=node07) which gave me a lot of undergoals and 2 chunks with goal=0
The undergoal was considered and the normal chunks where replicated to the other cs's.

Because of the check-loop it now showed me the two files with the missing chunks (which where endangered chunks before because the only chunk copy is on that specific node07 cs server)
The first chunk belongs to an mp3 file and the
second chunk belongs to a 14TB vbk backup file

When this was finished I started the cs process again having a lot of overgoals that where deleted in some hours as expected.
So when there where no overgoals the 2 chunks where endangerd again. I then copied the mp3 file to .bak and replaced the original file (with 1 endangered chunk) with this copy - so for this file there are again 2 copies.

But I don't know what to do with the large file - currently I have not enought free space to make a copy of the 14TB and then delete the original file with the missing 1 chunk as I did on the mp3 file.

Is there a tool that does the same as I did with the check-loop so I can find out what files are envolved for the next time?
Do you have any other advise I can try?
Probably it does not get replicated because of the huge amount of chunks of this file?
There are over 198.000 chunks for this file - the one chunk with 1 copy and all others as expected with 2 copies:
`
chunk 16513: 00000000000251BD_00000001 / (id:151997 ver:1)
                copy 1: 192.168.64.221:9422 (status:VALID)
                copy 2: 192.168.64.227:9422 (status:VALID)
                
chunk 16514: 00000000000251BE_00000001 / (id:151998 ver:1)
                copy 1: 192.168.64.227:9422 (status:VALID)
                
chunk 16515: 00000000000251BF_00000001 / (id:151999 ver:1)
                copy 1: 192.168.64.222:9422 (status:VALID)
                copy 2: 192.168.64.225:9422 (status:VALID)
`

To answer your questions about CRC in log - on the other servers there are no such entries, only on the node07 but from 13.10.2022 (when I switched on the cluster since August) - nothing newer about "crc" or "checksum" since then

`
Oct 13 11:21:40 node07 mfschunkserver[591]: hdd_senddata: CRC not flushed - writing now

Oct 13 12:25:33 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 12:41:28 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 12:44:19 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 13:04:20 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 13:09:36 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 13:36:28 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 13:39:44 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 13:57:09 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 14:20:11 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 14:21:07 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 14:21:59 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 14:26:56 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 14:30:21 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 14:36:04 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 14:38:28 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 14:42:38 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 14:47:41 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 15:03:25 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 15:17:47 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 15:18:31 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 15:29:48 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 15:30:48 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 15:56:52 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 16:00:40 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 16:12:52 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 16:32:40 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 16:40:22 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 16:51:33 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 16:52:54 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 17:08:52 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 17:17:36 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 17:21:09 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 17:27:10 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 17:46:25 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 17:50:14 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 17:52:09 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 17:55:49 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 17:57:13 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 18:32:10 node07 mfschunkserver[611]: replicator: write status: CRC error

Oct 13 18:32:10 node07 mfschunkserver[611]: replicator: write status: CRC error
`

**Comment by chogata** [comment]
OK, so I have bad news for you. The fact that all the other servers don't have any CRC errors in logs, only node07, means there is a physical problem with node07 (or its network connection).

When a chunk is being replicated from server A to server B, it goes like this:
- server A reads the data and its CRC and checks if it matches; if CRC would not match the data on server A, you would get a message in the logs AND this chunk copy would be marked invalid,
- server A sends the data and CRC to server B,
- server B reads data and CRC, calculates its own copy of CRC from read data and checks if the received CRC matches the calculated CRC; if it does not match, you get a CRC error in the logs.

Now, from your master log I can see that all unsuccessful replication tries were from one of the other nodes always to node07. There are no CRC errors on other nodes, meaning all "server As" in the above scenario checked the data and CRC before sending and it was correct. Yet "server B", aka node07, finds CRC errors. This means either the sent data got corrupted (possible if you have a faulty network card in node07) or node07 itself for some reason corrupts data (faulty RAM for example). You need to diagnose the problem and you cannot trust in integrity of any chunks recently recorded to node07.

So it's a good thing you stopped node07.

What to do in case you have a file with one missing chunk, that is only available on node07 - you can zero the chunk using `mfsfilerepair` on the affected file. But you can also copy this chunk manually from node07 to some other node. How to do it correctly:
- pick another node and stop the chunkserver process on it (it goes into maintenance mode for a while, so replications won't start immediately)
- locate the copy of the missing chunk on node07's hard drive and copy it over to the stopped node; you need to copy it to one of the storage directories of MooseFS's, so if you use, for example, a disk mounted in /mnt/chunks, there will be directories from 00 to FF in /mnt/chunks, pick one at random and copy the chunk there
- **important step, do not forget it!!!** delete `.chunkdb` file from /mnt/chunks (or wherever you have your chunks stored on the stopped node)
- start the stopped node again.

Because you deleted the `.chunkdb` file, the stopped node after starting will scan the disk with chunks to find them all. It will also find and report to the master the missing chunk you copied from node07. Your file will not have a missing chunk anymore, UNLESS this chunk turns out to be faulty and will be marked invalid by the system. If that happens, stop this node again, use `mfschunktool -r` to repair it and start the system again. But if you are forced to repair this chunk, then the contents is not guaranteed. Repairing basically means re-calculating CRC so it matches the data. If it is the CRC that got corrupted, then you will have correct data. If it's the data that got corrupted, well...

**Comment by OE2WNL** [comment]
Hi!

Thanks for your in depth replies!

I have some good news - I stopped node07 and copied - as you suggested - the one chunk, and I think this worked - as I now have no goal=0 files. Thank you!
(smartctl shows lot of UDMA CRC errors - so I would blame the HDD)

After this I replaced the disk in node07 and started the cs process.
Due to shutting down node07 there are a lot of endangered chunks.

Is it normal that filling the new disk of node07 has higher priority than replicating the endangered chunks ? Because since hours the endangered count stays at the same number (464016) but I can see that the rebalancing slowly happens.

I played with the settings on master as follows:
CHUNKS_WRITE_REP_LIMIT = 8,6,4,6
CHUNKS_READ_REP_LIMIT = 20,15,10,15

But the endangered count does not get lower. As with "top" I can see that node07 now has a load of 6 and "wait on disc" of 17% which means it's busy - but the other nodes are not doing much - why don't they replicate the endangered chunks?

I took a screenshot of the current status
"Bildschirmfoto_2022-10-20_19-53-37.png" and "Bildschirmfoto_2022-10-20_19-57-20.png"
https://share.derwebwolf.net/s/iLom03gBC8Y1wma

Goals are 2AB and 
node01-04 has Label A
node04-08 has Label B

There is space available on both groups but replication (of the endangered chunks) does not happen.

Thank you again for your time!

Greetings from austria.

-Wolfgang

**Comment by chogata** [comment]
Can you make screenshots of your Resources and Operations tabs?

**Comment by OE2WNL** [comment]
Please see:
Bildschirmfoto_2022-10-21_13-01-12.png
on https://share.derwebwolf.net/s/iLom03gBC8Y1wma

**Comment by OE2WNL** [comment]
out of nowhere at midnight the cs's started to reduce the endangered count - don't know why this started now but it is decreasing
I have uploaded a screenshot "Bildschirmfoto_2022-10-22_07-05-23.png"

**the log of node07 (with new hdd):**
Oct 21 19:04:01 node07 mfschunkserver[1766]: replicator: write status: CRC error
Oct 21 20:32:38 node07 mfschunkserver[1766]: replicator: write status: CRC error
Oct 21 21:25:05 node07 mfschunkserver[1766]: replicator: write status: CRC error
Oct 21 22:24:39 node07 mfschunkserver[1766]: replicator: write status: CRC error
Oct 22 00:05:00 node07 mfschunkserver[1766]: workers: 10+
Oct 22 00:05:00 node07 mfschunkserver[1766]: workers: 20+
Oct 22 00:05:00 node07 mfschunkserver[1766]: workers: 30+
Oct 22 00:12:03 node07 mfschunkserver[1766]: replicator: receive timed out
Oct 22 00:22:56 node07 mfschunkserver[1766]: replicator: write status: CRC error
Oct 22 00:30:49 node07 mfschunkserver[1766]: replicator: receive timed out

so the crc errors speaks for further problems (i.e. ram, sata-controller) on this node - right?

but exactly on 5 minutes past midnight all nodes started lot's of workers.
(the lot of worker might be caused by master settings I tried
CHUNKS_WRITE_REP_LIMIT = 30,30,5,5
CHUNKS_READ_REP_LIMIT = 50,50,5,5
)

**logs from master:**
Oct 21 22:24:39 mfsmaster mfsmaster[515]: (192.168.64.226:9422 -> 192.168.64.227:9422) chunk: 0000000000042A82 replication status: CRC error
Oct 21 23:00:00 mfsmaster mfsmaster[515]: child finished
Oct 21 23:00:00 mfsmaster mfsmaster[515]: store process has finished - store time: 0.971
Oct 21 23:15:04 mfsmaster mfsmaster[515]: structure check loop
Oct 22 00:00:00 mfsmaster mfsmaster[515]: child finished
Oct 22 00:00:00 mfsmaster mfsmaster[515]: store process has finished - store time: 0.930
Oct 22 00:06:38 mfsmaster mfsmaster[515]: (192.168.64.222:9422 -> 192.168.64.225:9422) chunk: 00000000000249C9 replication status: Disconnected
Oct 22 00:08:46 mfsmaster mfsmaster[515]: (192.168.64.223:9422 -> 192.168.64.225:9422) chunk: 0000000000264CB1 replication status: Disconnected
Oct 22 00:09:40 mfsmaster mfsmaster[515]: structure check loop
Oct 22 00:09:46 mfsmaster mfsmaster[515]: chunk 000000000002EE72_00000001: chunk hasn't been replicated since previous loop - retry
Oct 22 00:10:23 mfsmaster mfsmaster[515]: chunk 000000000004C5F5_00000001: chunk hasn't been replicated since previous loop - retry
Oct 22 00:10:25 mfsmaster mfsmaster[515]: chunk 0000000000036DFE_00000001: chunk hasn't been replicated since previous loop - retry
Oct 22 00:10:46 mfsmaster mfsmaster[515]: chunk 000000000022F818_00000001: chunk hasn't been replicated since previous loop - retry
Oct 22 00:11:17 mfsmaster mfsmaster[515]: chunk 000000000004C228_00000001: chunk hasn't been replicated since previous loop - retry
Oct 22 00:11:36 mfsmaster mfsmaster[515]: got replication status from server which had had that chunk before (chunk:000000000004C228_00000001)
...
...

nothing special at 00:05 in the log when replication started
what was holding them back - why exactly now ?

**Comment by OE2WNL** [comment]
Current state: now I have no endangered chunks, all have goal 2AB, system is now just balancing to bring the former empty hdd from node07 to same filling level.

All is good now - just wondering why the system waited with endangered-replicating:
* does it wait until the empty disc reached any special filling level?
* is there some kind of other waiting period?

**Comment by chogata** [comment]
Hi,
No, there is no special waiting period, apart from `REPLICATIONS_DELAY_INIT` in `mfsmaster.cfg`, but the default for that is 60 seconds and I assume if you set that to a hugely larger value, you would have remembered it ;) Also, chunks are not replicated when they are locked (that's why I was asking for you operations tab, to see if there is a lot if i/o that could potentially block replications, but there is none). MooseFS always has a reason, but this time I'm afraid the reason was an effect of some issues with node07.

The fact, that you still have CRC errors on node07 and only node07 means there is something wrong with the machine. Yes, RAM and/or sata controller need to be looked at (considering your previous disk gave you some grave errors, I would start with sata controller). This means data on node07 cannot be trusted, so whatever you have there is, in fact, endangered. I would recommend you to shut down this node for now and try to replace the faulty hardware, then wipe it and let the system rebalance again.

**Comment by OE2WNL** [comment]
Hi!
No the paramter REPLICATIONS_DELAY_INIT was not changed. (would be a nice explanation for my case but no ;-) )

I now have stopped cs on node07, let the endangered chunks replicate till all are replicated and no missing chunk is reported - and then?

As the cs nodes are simple Odroid HC2 (single chip computer) - I will replaced the whole node for a new one and format the drive and start the cs process (on new hardware) again? Or do I have to delete the cs id in `/var/lib/mfs/chunkserverid.mfs`also?

Thank you
-Wolfgang

**Comment by OE2WNL** [comment]
Hi!

Don't bother my previous question - I already replaced the node07 with new hardware, formatted the hdd and started the cs process.
It already replicated all chunks and I'm on track again.

So finally thank you again very much for your detailed answers to my questions/problems!
Have a nice day!
-Wolfgang

**Comment by OE2WNL** [comment]
As re-reading this discussion - one question remains:

Is there any tool where I can see to what file a special chunk belongs?
There has to be a way because when a file has missing chunks it is written at the cgi what file and what chunk it is about.
Thanks

**Comment by jkiebzak** [comment]
You can use `mfsfileinfo` to display what chunks belong to the file. But I
don't know of a way to display what file uses a specific chunk.

In the CGI, the list of files with missing chunks is created (I believe) by
a process that crawls all the files and confirming their chunk status. But
I could be wrong.

**Comment by chogata** [comment]
@jkiebzak is right, the list in CGI is built during a loop that checks files in the system and (among other things) shows missing files, if there are any. The list is not dynamic, it needs another loop run to update (if in the meantime any of the missing chunks were restored or some of the missing files were deleted by user).
